res/controllers/Behringer-CMDStudio4a-scripts.js

```diff
@@ -0,0 +1,257 @@
+
+
+// ************************ Initialisation stuff. *****************************
+
+BehringerCMDStudio4a.vuMeterUpdate = function (value, group, control){
+    value=(value*15)+48;
+    switch(control) {
+    case "VuMeterL":
+        midi.sendShortMsg(0xB0, 0x7E, value);
+        break;
+    case "VuMeterR":
+        midi.sendShortMsg(0xB0, 0x7F, value);
+        break;
+    }
+}
+
+BehringerCMDStudio4a.initLEDs = function () {
+    // Initialise any LEDs that are direcctly controlled by this script.
+    // Del buttons (one for each virtual deck).
+    midi.sendShortMsg(0x90, 0x2A, 0x00);
+    midi.sendShortMsg(0x91, 0x4A, 0x00);
+    midi.sendShortMsg(0x92, 0x2A, 0x00);
+    midi.sendShortMsg(0x93, 0x4A, 0x00);
+    // Scratch buttons (one for each virtual deck).
+    midi.sendShortMsg(0x90, 0x16, 0x00);
+    midi.sendShortMsg(0x91, 0x36, 0x00);
+    midi.sendShortMsg(0x92, 0x16, 0x00);
+    midi.sendShortMsg(0x93, 0x36, 0x00);    
+}
+
+BehringerCMDStudio4a.init = function () {
+    // Initialise anything that might not be in the correct state.
+    BehringerCMDStudio4a.initLEDs();
+    // Connect the VUMeters
+    engine.connectControl("[Master]","VuMeterL","BehringerCMDStudio4a.vuMeterUpdate");
+    engine.connectControl("[Master]","VuMeterR","BehringerCMDStudio4a.vuMeterUpdate");
+}
+ 
+BehringerCMDStudio4a.shutdown = function() {
+    // Leave the deck in a properly initialised state.
+    BehringerCMDStudio4a.initLEDs();
+
+    // Disconnect the VUMeters.
+    // Maybe not! This doesn't seem to work here!
+//  engine.connectControl("[Master]","VuMeterL","BehringerCMDStudio4a.vuMeterUpdate",true);
+//  engine.connectControl("[Master]","VuMeterR","BehringerCMDStudio4a.vuMeterUpdate",true);
+}
+
+
+// *************************** Control Stuff. *********************************
+// The code below is primarily "shift/mode" key functionality as there is no
+// native support for this in Mixxx at the moment.
+// I suspect that the vast majority of controller mappings could be completed
+// with little or no scripting if Mixxx supported shift/mode buttons in the
+// XML (together with standard wheel/scratching functionality).
+
+// Function to deal with the DEL "shift" buttons.
+BehringerCMDStudio4a.del = function (channel, control, value, status, group) {
+    BehringerCMDStudio4a.delState[channel]=!BehringerCMDStudio4a.delState[channel];
+    midi.sendShortMsg(status, control, BehringerCMDStudio4a.delState[channel] ? 0x01 : 0x00);
+}
+
+// Functions to deal with the play button.
+BehringerCMDStudio4a.playPush = function (channel, control, value, status, group) {
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // DEL is active, reverse-roll (slip)!
+        engine.setValue(group, "reverseroll", 1);
+    } else {
+        // No DEL, just play.
+        if (engine.getValue(group, "play")==0) {
+            engine.setValue(group, "play", 1);
+        } else {
+            engine.setValue(group, "play", 0);
+        }

```

Comment:
Ahh! OK, didn't know about the toggleControl() function. I've replaced the conditional toggle with that function now.

However, the push and release functions do different things depending on the mode, i.e. "toggle play" in normal mode, but "hold reverse-play" in  DELETE-mode, (and in fact the release function only has a defined behaviour in DELETE-mode), so I'm not convinced it's a simplification to use the same function for both (then have to add an extra if condition in the "super-function" to determine if the button has been pressed or released).

We have the ability to call different functions for press and release in the XML when the function behaviour for press and release is different, so we should use that ability to simplify the code right?

"Shift/mode" functionality is so fundamental to many controller mappings it seems odd that there is no native support for this in Mixxx at the moment, (i.e. support where you can just define the shift/mode behaviour in the XML with no additional scripting).